File: src/main/kotlin/com/github/npetrangelo/manimplugin/ManimSettingsEditor.java
package com.github.npetrangelo.manimplugin;

import com.intellij.openapi.options.ConfigurationException;
import com.intellij.openapi.options.SettingsEditor;
import com.intellij.openapi.ui.ComponentWithBrowseButton;
import com.intellij.openapi.ui.LabeledComponent;
import com.intellij.openapi.ui.TextFieldWithBrowseButton;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;

public class ManimSettingsEditor extends SettingsEditor<ManimRunConfiguration> {
    private JPanel myPanel;
    private LabeledComponent<ComponentWithBrowseButton> myMainClass;

    @Override
    protected void resetEditorFrom(@NotNull ManimRunConfiguration s) {

    }

    @Override
    protected void applyEditorTo(@NotNull ManimRunConfiguration s) throws ConfigurationException {

    }

    @Override
    protected @NotNull JComponent createEditor() {
        return myPanel;
    }

    private void createUIComponents() {
        myMainClass = new LabeledComponent<ComponentWithBrowseButton>();
        myMainClass.setComponent(new TextFieldWithBrowseButton());
    }
}

File: src/main/kotlin/com/github/npetrangelo/manimplugin/ManimSettingsEditor.kt
package com.github.npetrangelo.manimplugin

import com.intellij.openapi.fileChooser.FileChooserDescriptorFactory
import com.intellij.openapi.options.SettingsEditor
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.TextFieldWithBrowseButton
import com.intellij.ui.components.fields.ExpandableTextField
import com.intellij.ui.layout.panel
import javax.swing.JComponent
import javax.swing.JLabel

class ManimSettingsEditor(private val project : Project) : SettingsEditor<ManimConfig>() {
    private val fileLabel = JLabel("File:")
    private val fileText = TextFieldWithBrowseButton().apply {
        addBrowseFolderListener(
            "Select file",
            "Select the file that contains the scene you want to render.",
            project,
            FileChooserDescriptorFactory.createSingleFileDescriptor(".py")
        )
    }
    private val sceneLabel = JLabel("Scene:")
    private val sceneText = ExpandableTextField()

    override fun resetEditorFrom(m: ManimConfig) {
        fileText.text = m.runConfigSettings.filePath
        sceneText.text = m.runConfigSettings.scene
    }

    override fun applyEditorTo(m: ManimConfig) {
        m.runConfigSettings.filePath = fileText.text
        m.runConfigSettings.scene = sceneText.text
    }

    override fun createEditor(): JComponent = panel {
        row { fileLabel; fileText; }
        row { sceneLabel; sceneText; }
    }
}
